### PR TITLE
Ignore unlisted packages

### PR DESCRIPTION
**Pull request by troydai**
https://github.com/aspnet/XRE/issues/1208

### CODE REVIEW

**Comment by troydai** [comment]
@danroth27 @davidfowl @jeffhandley

File: src/Microsoft.Framework.PackageManager/Restore/RestoreCommand.cs [hunk 1]
```diff
@@ -30,7 +30,6 @@ public RestoreCommand(IApplicationEnvironment env)
         }
```
Comment (davidfowl) [inline src/Microsoft.Framework.PackageManager/Restore/RestoreCommand.cs]: Can you not format unrelated files :smile: ?

File: src/Microsoft.Framework.PackageManager/Restore/NuGet/NuGetv2Feed.cs [hunk 2]
```diff
@@ -21,6 +21,10 @@ public class NuGetv2Feed : IPackageFeed
         private static readonly XName _xnameProperties = XName.Get("properties", "http://schemas.microsoft.com/ado/2007/08/dataservices/metadata");
         private static readonly XName _xnameId = XName.Get("Id", "http://schemas.microsoft.com/ado/2007/08/dataservices");
         private static readonly XName _xnameVersion = XName.Get("Version", "http://schemas.microsoft.com/ado/2007/08/dataservices");
+        private static readonly XName _xnamePublish = XName.Get("Published", "http://schemas.microsoft.com/ado/2007/08/dataservices");
+
+        // An unlisted package's publish time mush be 1900-01-01T00:00:00.
+        private static readonly DateTime _unlistedPublishedTime = DateTime.Parse("1900-01-01T00:00:00");
```
Comment (davidfowl) [inline src/Microsoft.Framework.PackageManager/Restore/NuGet/NuGetv2Feed.cs]: Why are you parsing this instead of just newing it up?

File: src/Microsoft.Framework.PackageManager/Restore/NuGet/NuGetv2Feed.cs [hunk 3]
```diff
@@ -170,15 +175,24 @@ public PackageInfo BuildModel(string id, XElement element)
...
+                {
+                    // If 'Id' element exist, use its value as accurate package Id
+                    // Otherwise, use the value of 'title' if it exist
+                    // Use the given Id as final fallback if all elements above don't exist
+                    Id = idElement?.Value ?? titleElement?.Value ?? id,
+                    Version = SemanticVersion.Parse(properties.Element(_xnameVersion).Value),
+                    ContentUri = element.Element(_xnameContent).Attribute("src").Value,
+                };
+            }
+            else
```
Comment (davidfowl) [inline src/Microsoft.Framework.PackageManager/Restore/NuGet/NuGetv2Feed.cs]: Remove the else

File: src/Microsoft.Framework.PackageManager/Restore/NuGet/NuGetv2Feed.cs [hunk 4]
```diff
@@ -170,15 +175,24 @@ public PackageInfo BuildModel(string id, XElement element)
             var idElement = properties.Element(_xnameId);
             var titleElement = element.Element(_xnameTitle);
 
-            return new PackageInfo
+            var isUnlisted = DateTime.Parse(properties.Element(_xnamePublish).Value) == _unlistedPublishedTime;
+
+            if (!isUnlisted)
```
Comment (davidfowl) [inline src/Microsoft.Framework.PackageManager/Restore/NuGet/NuGetv2Feed.cs]: invert the logic, it makes it clearer.

File: src/Microsoft.Framework.PackageManager/Restore/NuGet/NuGetv2Feed.cs [hunk 5]
```diff
@@ -170,15 +175,24 @@ public PackageInfo BuildModel(string id, XElement element)
             var idElement = properties.Element(_xnameId);
             var titleElement = element.Element(_xnameTitle);
 
-            return new PackageInfo
+            var isUnlisted = DateTime.Parse(properties.Element(_xnamePublish).Value) == _unlistedPublishedTime;
```
Comment (davidfowl) [inline src/Microsoft.Framework.PackageManager/Restore/NuGet/NuGetv2Feed.cs]: Code defensively here. Use TryParse and verify that the value isn't null.

File: src/Microsoft.Framework.PackageManager/Restore/RestoreCommand.cs (hunk 1 above)
Comment (troydai) [inline src/Microsoft.Framework.PackageManager/Restore/RestoreCommand.cs]: I ... just ... can't ... help ... doing ... that ...

Comment (davidfowl) [inline src/Microsoft.Framework.PackageManager/Restore/RestoreCommand.cs]: Do it in a stand alone pull request after.

**Comment by troydai** [comment]
@jeffhandley @davidfowl 
A few questions:
1. Can `d:Published` missing?
2. In the case if it is missing, can I regard this package as listed?

**Comment by troydai** [comment]
Ping @davidfowl @danroth27 @jeffhandley

File: src/Microsoft.Framework.PackageManager/Restore/NuGet/NuGetv2Feed.cs [hunk 6]
```diff
@@ -21,6 +21,10 @@ public class NuGetv2Feed : IPackageFeed
         private static readonly XName _xnameProperties = XName.Get("properties", "http://schemas.microsoft.com/ado/2007/08/dataservices/metadata");
         private static readonly XName _xnameId = XName.Get("Id", "http://schemas.microsoft.com/ado/2007/08/dataservices");
         private static readonly XName _xnameVersion = XName.Get("Version", "http://schemas.microsoft.com/ado/2007/08/dataservices");
+        private static readonly XName _xnamePublish = XName.Get("Published", "http://schemas.microsoft.com/ado/2007/08/dataservices");
+
+        // An unlisted package's publish time mush be 1900-01-01T00:00:00.
```
Comment (BrennanConroy) [inline src/Microsoft.Framework.PackageManager/Restore/NuGet/NuGetv2Feed.cs]: mush -> must

File: src/Microsoft.Framework.PackageManager/Restore/NuGet/NuGetv2Feed.cs [hunk 7]
```diff
@@ -170,6 +175,23 @@ public PackageInfo BuildModel(string id, XElement element)
             var idElement = properties.Element(_xnameId);
             var titleElement = element.Element(_xnameTitle);
```
Comment (BrennanConroy) [inline src/Microsoft.Framework.PackageManager/Restore/NuGet/NuGetv2Feed.cs]: Move these below the publishElement checks since they aren't used until after the checks?

**Comment by jeffhandley** [comment]
I don't think Published can be missing, but it can't hurt to guard against it in case someone's custom NuGet feed doesn't have it.  I would interpret no Published field to Listed, as that would mean they cannot differentiate between listed and unlisted.

**Comment by davidfowl** [comment]
+1 @jeffhandley Make the change @troydai and :shipit:

**Comment by troydai** [comment]
Thanks for the advice. I'll update accordingly.

File: src/Microsoft.Framework.PackageManager/Restore/NuGet/NuGetv2Feed.cs [hunk 8]
```diff
@@ -170,6 +175,16 @@ public PackageInfo BuildModel(string id, XElement element)
...
 
+            var publishElement = properties.Element(_xnamePublish);
+            if (publishElement != null)
+            {
+                DateTime publishDate; 
+                if (DateTime.TryParse(publishElement.Value, out publishDate) && (publishDate == _unlistedPublishedTime))
```
Comment (victorhurdugaci) [inline src/Microsoft.Framework.PackageManager/Restore/NuGet/NuGetv2Feed.cs]: Can you make `IsPublished` a property of the `PackageInfo` class? It might be useful for other purposes, not just for this check

File: src/Microsoft.Framework.PackageManager/Restore/NuGet/NuGetv2Feed.cs [hunk 9]
```diff
@@ -170,6 +175,16 @@ public PackageInfo BuildModel(string id, XElement element)
...
+                {
+                    return null;
```
Comment (victorhurdugaci) [inline src/Microsoft.Framework.PackageManager/Restore/NuGet/NuGetv2Feed.cs]: If you do what I said above, then you can return the model always and then only select those that are published

Comment (troydai) [inline src/Microsoft.Framework.PackageManager/Restore/NuGet/NuGetv2Feed.cs]: I'm not sure about changing the `PackageInfo` interface. The `IsPublish` value maybe useful, but we don't know yet. We can get back to save the value once we have a case.
